### PR TITLE
Add Package Instances #2691

### LINKED ISSUE #2691: Package model updates: Add new `package_instance`
Package model updates: Add new `package_instance` attribute used to track unique instances of each packages (say you have two identical wheels in the same code tree);use a UUID for this `package_instance` attribute.

### PR DESCRIPTION
**Pull request by AyanSinhaMahapatra**
This PR adds the PackageInstance class and functions to group package
manifests and package data as top level package instances.

Existing package data are ported to this new approach.

Reference: https://github.com/nexB/scancode-toolkit/issues/2098 
Reference: https://github.com/nexB/scancode-toolkit/issues/2691 
Reference: https://github.com/nexB/scancode-toolkit/issues/2692
Reference: https://github.com/nexB/scancode-toolkit/issues/2693
Reference: https://github.com/nexB/scancode-toolkit/issues/2843 
Reference: https://github.com/nexB/scancode-toolkit/issues/2652 
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

### CODE REVIEW

File: tests/packagedcode/test_package_instance.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+class TestPackageInstance(PackageTester):
+    test_data_dir = os.path.join(os.path.dirname(__file__), 'data')
+
+    def test_package_instance_scan_python(self):
+        test_dir = self.get_test_loc('instance/pypi')
+        result_file = self.get_temp_file('json')
+        expected_file = self.get_test_loc('instance/python-package-instance-expected.json')
+        run_scan_click(['--package', '--strip-root', '--processes', '-1', test_dir, '--json', result_file])
+        check_json_scan(expected_file, result_file, regen=True)
```
Comment (JonoYang) [inline tests/packagedcode/test_package_instance.py]: `regen` should be set to `False`

File: src/packagedcode/__init__.py [hunk 2]
```diff
@@ -110,28 +110,36 @@
...
-            raise Exception(msg)
-        else:
-            seen_types[manifest.package_manifest_type] = manifest
+
+PACKAGE_INSTANCES_BY_TYPE = {
+    cls.default_type: cls
+    for cls in PACKAGE_INSTANCE_TYPES
+}
+
+def check_package_manifest_classes():
```
Comment (JonoYang) [inline src/packagedcode/__init__.py]: Is `check_package_manifest_classes()` called anywhere? Previously, the code in the function was outside of a function. My assumption was that whenever the `packagedcode` module is imported, this code that checks `PACKAGE_MANIFESTS_BY_TYPE` and `PACKAGE_MANIFEST_TYPES` would always run and give us an exception whenever a package was not properly set.

File: src/packagedcode/__init__.py [hunk 3]
```diff
@@ -110,28 +110,36 @@
...
-                  .format(manifest.package_manifest_type, manifest.__name__, seen.package_manifest_type, seen.__name__,))
-            raise Exception(msg)
-        else:
-            seen_types[manifest.package_manifest_type] = manifest
+
+PACKAGE_INSTANCES_BY_TYPE = {
+    cls.default_type: cls
+    for cls in PACKAGE_INSTANCE_TYPES
+}
+
```
Comment (JonoYang) [inline src/packagedcode/__init__.py:150]: Style nitpick: there should be an extra new line here.

File: src/packagedcode/models.py [hunk 4]
```diff
@@ -765,6 +765,251 @@ def recognize(cls, location):
...
+        Returns a mapping of package data attributes and corresponding values.
+        """
+        mapping = self.to_dict()
+
+        # Removes PackageInstance specific attributes
+        for attribute in ('package_uuid', 'package_manifest_paths', 'files'):
+            mapping.pop(attribute)
+        
+        # Remove attributes which are BasePackage functions
+        for attribute in ('repository_homepage_url', 'repository_download_url', 'api_data_url'):
```
Comment (JonoYang) [inline src/packagedcode/models.py:832]: I have no strong feelings one way or the other about this, but I am curious as to why `repository_homepage_url`, `repository_download_url`, and `api_data_url` are removed from the package data that is to be returned.

**Review by JonoYang — COMMENTED** [review]
@AyanSinhaMahapatra Initial work on the `PackageInstance` model looks good. I just have some minor things I pointed out.

File: tests/packagedcode/test_package_instance.py (hunk 1 above)
Comment (AyanSinhaMahapatra) [reply tests/packagedcode/test_package_instance.py]: I was having some problems with the `package_uuid` field as it had a UUID that was different for each test run. So I forgot to make this False, now this is resolved. I've also modified the `check_json_scan` to accept a new argument `remove_package_uuid` which removes the UUID qualifier from the `package_uuid` pURL string.

File: src/packagedcode/__init__.py (hunk 2 above)
Comment (AyanSinhaMahapatra) [reply src/packagedcode/__init__.py]: See [here](https://github.com/nexB/scancode-toolkit/pull/2748#discussion_r761111416) this was supposed to be moved into a test instead, but it was not completed. Now I've pushed the proper tests.

File: src/packagedcode/__init__.py (hunk 3 above)
Comment (AyanSinhaMahapatra) [reply src/packagedcode/__init__.py:150]: Added, thanks!

Comment (AyanSinhaMahapatra) [reply src/packagedcode/__init__.py:150]: Added, thanks!

File: src/packagedcode/models.py (hunk 4 above)
Comment (AyanSinhaMahapatra) [reply src/packagedcode/models.py:832]: So this is a helper function used in the merging of package_manifest data into a package_instance [here](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L897). These are basically fields which are not set as values in a package data, but obtained from functions like [this](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L320) in the `to_dict()`. So we don't need to merge these data from the mappings.

**Comment by pombredanne** [comment]
re: https://github.com/nexB/scancode-toolkit/pull/2825#discussion_r800437754

> I was having some problems with the package_uuid field as it had a UUID that was different for each test run. So I forgot to make this False, now this is resolved. I've also modified the check_json_scan to accept a new argument remove_package_uuid which removes the UUID qualifier from the package_uuid pURL string.


There are three ways to handle these cases:

1. ignore it as you have done but you need to have at least one test case that would not ignore the UUID to ensure this is generated correctly.
2. use mocks in tests so that you have a stable mocked UUID created and do not need to ignore it
3. add something like an extra `_test=True` argument to the tested function(s) that is called only in tests and that creates a static UUID that is well known for testing.

The more generic solution is 2. ... use mock.

**Comment by AyanSinhaMahapatra** [comment]
@pombredanne This is ready!

**Review by pombredanne — APPROVED** [review]
LGTM. Thanks! merging now.

**Pull request merged**